libnavui-maps/src/test/java/com/mapbox/navigation/ui/maps/internal/route/line/MapboxRouteLineUtilsTest.kt

```diff
@@ -2063,17 +2084,216 @@ class MapboxRouteLineUtilsTest {
         verify { style.removeStyleLayer(LAYER_GROUP_3_RESTRICTED) }
     }
 
-    private fun <T> listElementsAreEqual(
-        first: List<T>,
-        second: List<T>,
-        equalityFun: (T, T) -> Boolean
-    ): Boolean {
-        if (first.size != second.size) {
-            return false
+    @Test
+    fun cacheResultKeyRouteTraffic_hashCode_legacyTraffic() {
+        val route = loadNavigationRoute("short_route.json")
+        val trafficProvider = MapboxRouteLineUtils.getRouteLegTrafficCongestionProvider
+        var expectedHashCode = route.id.hashCode()
+        expectedHashCode = 31 * expectedHashCode + trafficProvider.hashCode()
+        route.directionsRoute.legs()?.forEach { routeLeg ->
+            expectedHashCode =
+                31 * expectedHashCode + routeLeg.annotation()?.congestion().hashCode()
+            expectedHashCode =
+                31 * expectedHashCode + routeLeg.annotation()?.congestionNumeric().hashCode()
+            expectedHashCode = 31 * expectedHashCode + routeLeg.closures().hashCode()
+        }
+
+        val hashCode = CacheResultUtils.CacheResultKeyRouteTraffic<Any>(
+            route,
+            trafficProvider
+        ).hashCode()
+
+        assertEquals(expectedHashCode, hashCode)
+    }
+
+    @Test
+    fun cacheResultKeyRouteTraffic_hashCode_multiLegRoute() {
+        val route = loadNavigationRoute("multileg-route-two-legs.json")
+        val trafficProvider =
+            MapboxRouteLineUtils.getRouteLegTrafficCongestionProvider
+        var expectedHashCode = route.id.hashCode()
+        expectedHashCode = 31 * expectedHashCode + trafficProvider.hashCode()
+        route.directionsRoute.legs()?.forEach { routeLeg ->
+            expectedHashCode =
+                31 * expectedHashCode + routeLeg.annotation()?.congestion().hashCode()
+            expectedHashCode =
+                31 * expectedHashCode + routeLeg.annotation()?.congestionNumeric().hashCode()
+            expectedHashCode = 31 * expectedHashCode + routeLeg.closures().hashCode()
+            MapboxRouteLineUtils.getRoadClassArray(routeLeg.steps()).forEach {
+                expectedHashCode = 31 * expectedHashCode + it.hashCode()
+            }
+        }
+
+        val hashCode = CacheResultUtils.CacheResultKeyRouteTraffic<Any>(
+            route,
+            trafficProvider
+        ).hashCode()
+
+        assertEquals(expectedHashCode, hashCode)
+    }
+
+    @Test
+    fun cacheResultKeyRouteTraffic_equals() {
```

Comment:
You can also check:
1) different legs size;
2) different route ids;

Reply:
Explicit tests added.